TF-M: regular upmerge, post Zephyr v2.6.0

ioannisg opened the pull request:
- TF-M upmerge with several fixes. 
  upmerge up to upstream commit SHA: a199c3047f320a2f82b9a0c27af5b50991184e0f
  In addition to the fixes, the upmerge brings in the new FLIH service.
- We remove the TF-M specific module for mcuboot. TF-M builds in Zephyr will now use the mcuboot that Zephyr comes with.

ioannisg committed:
modules: tf-m: update target names for arm targets

Upstream TF-M has modified the path name for ARM platforms,
so we fix the board names in the module integration code.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
manifest: update trusted-firmware-m module SHA

Update TF-M module SHA to pick up the latest version
of the upstream project.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: stm: switch postbuild.sh script to a command from a byproduct

Running postbuild.sh script after TF-M builds is a command,
to be executed after build, not a post-build byproduct.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
modules: tf-m: supply NXP HAL directory path for builds for LPC

For TF-M builds on LPCXPRESSO55S69 we set the path to the
NXP HAL sources that are required for building TF-M for this
platform.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
samples: tfm_ipc: call sys_arch_reboot(..) directly

Switch to calling the sys_arch_reboot() API directly,
instead of going through the generic sys_reboot API.
This is to avoid locking the IRQs before the reset is
called, which is breaking the PSA call execution.

Also, align sample code to use the same secure service
(platform reset) to show case the PSA connect and close
APIs, instead of a random one (since the reset service
has to work to ultimately perform the system reset).

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

microbuilder commented:
Regression tests are failing for PS in QEMU, but this seems to be an issue with either QEMU or the upstream TF-M project, not with Zephyr.

The issue can be reproduce using only upstream TF-M by cloning TF-M and running the following script. It has been tested on QEMU 6.0.0 (by myself) andd 5.1.0 (Ioannis):

```bash
#!/bin/bash

# Exit on error
set -e

# Run from 'trusted-firmware-m'
rm -rf cmake_build
cmake -S . -B cmake_build -DTFM_PLATFORM=arm/mps2/an521 -DTFM_TOOLCHAIN_FILE=toolchain_GNUARM.cmake -DCMAKE_BUILD_TYPE=Debug -DTEST_S=ON -DTEST_NS=ON
cmake --build cmake_build -- install

# Generate a merged binary
srec_cat cmake_build/install/outputs/ARM/MPS2/AN521/tfm_s_ns_signed.bin -Binary -offset=0x80000 -o tfm_s_ns_signed.hex -Intel
srec_cat cmake_build/install/outputs/ARM/MPS2/AN521/bl2.hex -Intel tfm_s_ns_signed.hex -Intel -o tfm_full.hex -Intel

# Optional: Run in qemu
qemu-system-arm -M mps2-an521 -device loader,file=tfm_full.hex -serial stdio
```

This results in the regression tests consistently hanging in the PS test suite:

```
*** End of Secure test suites ***

#### Execute test suites for the Non-secure area ####
Running Test Suite PSA protected storage NS interface tests (TFM_NS_PS_TEST_1XXX)...
> Executing 'TFM_NS_PS_TEST_1001' 
  Description: 'Set interface'
  TEST: TFM_NS_PS_TEST_1001 - PASSED!
> Executing 'TFM_NS_PS_TEST_1002' 
  Description: 'Set interface with create flags'
  TEST: TFM_NS_PS_TEST_1002 - PASSED!
> Executing 'TFM_NS_PS_TEST_1003' 
  Description: 'Set interface with NULL data pointer'
  TEST: TFM_NS_PS_TEST_1003 - PASSED!
> Executing 'TFM_NS_PS_TEST_1004' 
  Description: 'Set interface with write once UID'
  TEST: TFM_NS_PS_TEST_1004 - PASSED!
> Executing 'TFM_NS_PS_TEST_1005' 
  Description: 'Get interface with valid data'
  TEST: TFM_NS_PS_TEST_1005 - PASSED!
> Executing 'TFM_NS_PS_TEST_1006' 
  Description: 'Get interface with zero data length'
  TEST: TFM_NS_PS_TEST_1006 - PASSED!
> Executing 'TFM_NS_PS_TEST_1007' 
  Description: 'Get interface with invalid UIDs'
  TEST: TFM_NS_PS_TEST_1007 - PASSED!
> Executing 'TFM_NS_PS_TEST_1008' 
  Description: 'Get interface with invalid data lengths and offsets'
  TEST: TFM_NS_PS_TEST_1008 - PASSED!
> Executing 'TFM_NS_PS_TEST_1009' 
  Description: 'Get interface with NULL data pointer'
  TEST: TFM_NS_PS_TEST_1009 - PASSED!
> Executing 'TFM_NS_PS_TEST_1010' 
  Description: 'Get info interface with write once UID'
  TEST: TFM_NS_PS_TEST_1010 - PASSED!
> Executing 'TFM_NS_PS_TEST_1011' 
  Description: 'Get info interface with valid UID'
  TEST: TFM_NS_PS_TEST_1011 - PASSED!
> Executing 'TFM_NS_PS_TEST_1012' 
  Description: 'Get info interface with invalid UIDs'
  TEST: TFM_NS_PS_TEST_1012 - PASSED!
> Executing 'TFM_NS_PS_TEST_1013' 
  Description: 'Remove interface with valid UID'
```

ioannisg commented:
@microbuilder I will disable the test temporarily for mps2_an521_ns and file issues in Zephyr and TFM upstream.

ioannisg commented:
> @microbuilder I will disable the test temporarily for mps2_an521_ns and file issues in Zephyr and TFM upstream.

https://developer.trustedfirmware.org/T951

ioannisg committed:
samples: tfm_integration: disable regression tests sample for AN521

NS regression tests are hanging on qemu for AN521. The problem is
present in the upstream project as well (issue is filed) so disable
the tests for now on QEMU.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
manifest: remove mcuboot version for TF-M in Zephyr

We remove the tfm-mcuboot module in Zephyr (which hosts the
mcuboot version that is used for TF-M builds in Zephyr). We
will force TF-M builds in Zephyr to  use the MCUboot version
that any Zephyr builds are using. This will allow Zephyr to
maintain a single mcuboot module version in its default
manifest.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

microbuilder reviewed (approved):
LGTM

microbuilder commented:
Adding DNM until bootloader issues can be resolved.

ioannisg commented:
> Adding DNM until bootloader issues can be resolved.

@microbuilder I believe the issues here are unrelated to TF-M. I am not 100% sure why CI needs to run tests on platforms that do not use  TF-M...

microbuilder commented:
> > Adding DNM until bootloader issues can be resolved.
> 
> @microbuilder I believe the issues here are unrelated to TF-M. I am not 100% sure why CI needs to run tests on platforms that do not use TF-M...

@ioannisg The problem here is with MCUBoot, and the last minute update of MCUBoot in this PR, since there are now two identical references to the same library:

- https://github.com/zephyrproject-rtos/zephyr/blob/c078fa45b32e8cc4ea4d2264a1fd2e17a0ac6585/west.yml#L113
- https://github.com/zephyrproject-rtos/zephyr/blob/c078fa45b32e8cc4ea4d2264a1fd2e17a0ac6585/west.yml#L147

This leads to tthe `MCUBOOT_BOOTUTIL` library being added twice: https://github.com/zephyrproject-rtos/mcuboot/blob/main/boot/bootutil/zephyr/CMakeLists.txt#L7

ioannisg commented:
> > > Adding DNM until bootloader issues can be resolved.
> > 
> > 
> > @microbuilder I believe the issues here are unrelated to TF-M. I am not 100% sure why CI needs to run tests on platforms that do not use TF-M...
> 
> @ioannisg The problem here is with MCUBoot, and the last minute update of MCUBoot in this PR, since there are now two identical references to the same library:
> 
>     * https://github.com/zephyrproject-rtos/zephyr/blob/c078fa45b32e8cc4ea4d2264a1fd2e17a0ac6585/west.yml#L113
> 
>     * https://github.com/zephyrproject-rtos/zephyr/blob/c078fa45b32e8cc4ea4d2264a1fd2e17a0ac6585/west.yml#L147
> 
> 
> This leads to tthe `MCUBOOT_BOOTUTIL` library being added twice: https://github.com/zephyrproject-rtos/mcuboot/blob/main/boot/bootutil/zephyr/CMakeLists.txt#L7

Yes, I figured this out as well, you're correct. 
to elaborate a bit more:
- some recent MCUboot changes have made MCUBOOT_BOOTUTIL an interface library.
- when we upgrate tfm-mcuboot, we are bringing in this change
- as you point out, this library is added twice in the build, under some configurations

What are our options?
1. Do not allow the build system (cmake) to add the tfm-mcuboot subdirectory, for the current build. This makes sense because tfm-mcuboot is _not_ a Zephyr module; it is used only by the TFM build. We should have the possibility to exclude some "modules" in west from being added to the build. To the best of my knowledge, this is not supported today, but IMHO it should be supported, because we might want to exlcude some module paths from our builds in certain situations (such as this one)
2. Do not use a specific MCUboot version for TF-M in  Zephyr. This is the "right thing to do", in accordance with how we addressed the mbedTLS case: Zephyr and TF-M should have the same versions (if possible) for dependencies such as MCUboot or mbedtls (possibly others).

For the purpose of the current upmerge, I am going to try to go for option (2). But I will also file an issue about (1) for @tejlmand to consider, when he is back.

ioannisg committed:
modules: tf-m: add TFM_BL2 dependency on TFM_MCUBOOT_IMAGE_NUMBER

TFM_MCUBOOT_IMAGE_NUMBER Kconfig option should have
a dependency on TFM_BL2, since it is not applicable
otherwise.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
modules: tf-m: allow DOWNLOAD option for mcuboot

Allow developers the option of using the DOWNLOAD option
for MCUboot in TF-M bulds with BL2. Using MCUboot from
Zephyr is still the default option.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
tests: arm_thread_swap_tz: increase computational length for secure call

In some platforms, such as the nucleo L552ZE_Q, the hash computation
is done very fast, before the delayed work manages to trigger (and
perform the assert-based validations). Increase the length of the
computations so this race condition can be avoided.

Also, this commit adds some messages in the assert statements for
ease of debugging.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
samples: tfm_integration: adapt DTS overlays for STM32L562 board

Due to changes in flash_layout.h for STM32L562, in the current
TF-M module update, we need to modify the DTS overlay files in
the TF-M samples where the board is supported.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
samples: tfm_ipc: add test-variant for TF-M without BL2

We add a test variant in tfm_ipc test, to validate the
scenario where TF-M is built without BL2 (MCUboot). We
test this on QEMU only (MPS2 AN521).

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
manifest: mcuboot: update to fix error with integration platforms

Minor update to mcuboot to fix the issue with the integration
platforms and allow list in the zephyr boot sample.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

nvlsianpu commented on west.yml:
```diff
@@ -162,7 +162,7 @@ manifest:
         - crypto
     - name: mcuboot
-      revision: 7a5196820ba48a6ad7e7d573c9048c021960677c
+      revision: pull/55/head
```
you can update the SHA

ioannisg replied:
done now, thanks!
